File: Aula01/Codigos/Veiculo.py
class Veiculo:
    # __init_ => é o método construtor
    def __init__(self, marca, modelo, placa, ano):
        self.marca = marca
        self.modelo = modelo
        self.placa = placa
        self.ano = ano
    # Método de instância
    def calcularTempoUso(self):
        return 2024 - self.ano


# Abaixo estou instanciando um objeto
#   da Classe Veiculo
meuUno = Veiculo("Fiat", "Uno com Escada", "ABC-1234", 2005)
meuUno.ano = 2007
teuFusca = Veiculo("Volks", "Fusca do Itamar", "DEF-", 1995)

print(meuUno.calcularTempoUso())
print(teuFusca.calcularTempoUso())

print("")

File: Aula10/Codigos/CodigosEmAula/Veiculo.java
import java.time.LocalDate;

class Veiculo {
    String marca;
    String modelo;
    int ano;
    String placa;
    public Veiculo() {
        //Método construtor padrão
    }
    public Veiculo(String marca, String modelo, int ano, String placa) {
        this.marca = marca;
        this.modelo = modelo;
        this.ano = ano;
        this.placa = placa;
    }
    Veiculo cloneFromOther(Veiculo veiculoOrigem) {
        Veiculo novoVeiculo = new Veiculo();
        novoVeiculo.marca = veiculoOrigem.marca;
        novoVeiculo.modelo = veiculoOrigem.modelo;
        novoVeiculo.ano = veiculoOrigem.ano;
        novoVeiculo.placa = veiculoOrigem.placa;
        return novoVeiculo;
    }
    //Abaixo temos um exemplo de "clone" errado
    // Pois temos a atribuição por referência
    // Nesse caso o objeto que é recebido por parâmetro
    // é o mesmo retornado (return)
    Veiculo cloneFromOtherWrong(Veiculo veiculoOrigem) {
        Veiculo novoVeiculo = veiculoOrigem;
        return novoVeiculo;
    }

    Veiculo cloneMe() {
        Veiculo novoVeiculo = new Veiculo();
        novoVeiculo.marca = this.marca;
        novoVeiculo.modelo = this.modelo;
        novoVeiculo.ano = this.ano;
        novoVeiculo.placa = this.placa;
        return novoVeiculo;
    }


    int calcularTempoUso() {
        int anoAtual = LocalDate.now().getYear();
        return this.calcularTempoUso(anoAtual);
    }

    int calcularTempoUso(int anoBase) {
        int tempoUso = anoBase - this.ano;
        return tempoUso;
    }
}